File: src/main/java/com/nextfit/nutrition/service/ConsultationService.java
package com.nextfit.nutrition.service;

import com.nextfit.nutrition.model.ConsultationDieteticien;
import com.nextfit.nutrition.model.Utilisateur;
import com.nextfit.nutrition.repository.ConsultationRepository;
import com.nextfit.nutrition.repository.UtilisateurRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Optional;

/**
 * Service de gestion des consultations diététicien en ligne.
 * Les abonnés VIP bénéficient de consultations incluses,
 * les autres membres paient 45€ par séance.
 */
@Service
@Transactional
public class ConsultationService {

    private final ConsultationRepository consultationRepo;
    private final UtilisateurRepository utilisateurRepo;

    public ConsultationService(ConsultationRepository consultationRepo,
                                UtilisateurRepository utilisateurRepo) {
        this.consultationRepo = consultationRepo;
        this.utilisateurRepo = utilisateurRepo;
    }

    /**
     * Réserve une consultation pour un membre.
     * Vérifie que le créneau n'est pas déjà pris.
     */
    public ConsultationDieteticien reserverConsultation(Long utilisateurId, LocalDateTime dateHeure,
                                                         int dureeMins) {
        if (consultationRepo.existsByDateHeure(dateHeure)) {
            throw new IllegalStateException("Ce créneau est déjà réservé : " + dateHeure);
        }

        Utilisateur utilisateur = utilisateurRepo.findById(utilisateurId)
                .orElseThrow(() -> new IllegalArgumentException("Utilisateur introuvable : " + utilisateurId));

        ConsultationDieteticien consultation = new ConsultationDieteticien(utilisateur, dateHeure);
        consultation.setDureeMins(dureeMins);

        // Abonnés VIP : consultation offerte
        boolean estVip = utilisateur.getTypeAbonnement() == Utilisateur.TypeAbonnement.VIP;
        consultation.setIncluse(estVip);
        consultation.setPrixEuros(estVip ? 0.0 : 45.0);

        // Assigner un diététicien (logique simplifiée)
        consultation.setNomDieteticien("Dr. Marie Dupont - Diététicienne agréée");
        consultation.setLienVisio("https://visio.nextfit.fr/session/" + System.currentTimeMillis());
        consultation.setStatut(ConsultationDieteticien.StatutConsultation.CONFIRMEE);

        return consultationRepo.save(consultation);
    }

    /** Annule une consultation */
    public ConsultationDieteticien annulerConsultation(Long consultationId) {
        ConsultationDieteticien consultation = consultationRepo.findById(consultationId)
                .orElseThrow(() -> new IllegalArgumentException("Consultation introuvable : " + consultationId));

        if (consultation.getStatut() == ConsultationDieteticien.StatutConsultation.TERMINEE) {
            throw new IllegalStateException("Impossible d'annuler une consultation déjà terminée.");
        }

        consultation.setStatut(ConsultationDieteticien.StatutConsultation.ANNULEE);
        return consultationRepo.save(consultation);
    }

    /** Consultations d'un utilisateur */
    @Transactional(readOnly = true)
    public List<ConsultationDieteticien> consultationsUtilisateur(Long utilisateurId) {
        return consultationRepo.findByUtilisateurIdOrderByDateHeureDesc(utilisateurId);
    }

    /** Prochaines consultations d'un utilisateur */
    @Transactional(readOnly = true)
    public List<ConsultationDieteticien> prochainesConsultations(Long utilisateurId) {
        return consultationRepo.findByUtilisateurIdAndDateHeureAfterOrderByDateHeureAsc(
                utilisateurId, LocalDateTime.now());
    }

    @Transactional(readOnly = true)
    public Optional<ConsultationDieteticien> trouverParId(Long id) {
        return consultationRepo.findById(id);
    }
}

File: Services/ConsultationService.cs
using Microsoft.EntityFrameworkCore;
using NextfitNutrition.Data;
using NextfitNutrition.Models;

namespace NextfitNutrition.Services;

/// <summary>
/// Service de gestion des consultations diététicien en ligne.
/// Les abonnés VIP ont leurs consultations incluses.
/// </summary>
public class ConsultationService
{
    private readonly NutritionDbContext _ctx;

    public ConsultationService(NutritionDbContext ctx) => _ctx = ctx;

    /// <summary>Réserve un créneau. Vérifie que le créneau n'est pas déjà pris.</summary>
    public async Task<ConsultationDieteticien> ReserverAsync(long utilisateurId,
                                                               DateTime dateHeure,
                                                               int dureeMins = 30)
    {
        if (await _ctx.Consultations.AnyAsync(c => c.DateHeure == dateHeure))
            throw new InvalidOperationException($"Ce créneau est déjà réservé : {dateHeure:g}");

        var utilisateur = await _ctx.Utilisateurs.FindAsync(utilisateurId)
            ?? throw new ArgumentException($"Utilisateur introuvable : {utilisateurId}");

        bool estVip = utilisateur.TypeAbonnement == Utilisateur.TypeAbonnement.VIP;

        var consultation = new ConsultationDieteticien
        {
            UtilisateurId   = utilisateurId,
            DateHeure       = dateHeure,
            DureeMins       = dureeMins,
            Statut          = ConsultationDieteticien.StatutConsultation.CONFIRMEE,
            NomDieteticien  = "Dr. Marie Dupont — Diététicienne agréée",
            LienVisio       = $"https://visio.nextfit.fr/session/{DateTimeOffset.UtcNow.ToUnixTimeMilliseconds()}",
            Incluse         = estVip,
            PrixEuros       = estVip ? 0.0 : 45.0
        };

        _ctx.Consultations.Add(consultation);
        await _ctx.SaveChangesAsync();
        return consultation;
    }

    /// <summary>Annule une consultation.</summary>
    public async Task<ConsultationDieteticien> AnnulerAsync(long consultationId)
    {
        var consultation = await _ctx.Consultations.FindAsync(consultationId)
            ?? throw new ArgumentException($"Consultation introuvable : {consultationId}");

        if (consultation.Statut == ConsultationDieteticien.StatutConsultation.TERMINEE)
            throw new InvalidOperationException("Impossible d'annuler une consultation déjà terminée.");

        consultation.Statut = ConsultationDieteticien.StatutConsultation.ANNULEE;
        await _ctx.SaveChangesAsync();
        return consultation;
    }

    public async Task<List<ConsultationDieteticien>> GetParUtilisateurAsync(long utilisateurId)
        => await _ctx.Consultations
            .Where(c => c.UtilisateurId == utilisateurId)
            .OrderByDescending(c => c.DateHeure)
            .ToListAsync();

    public async Task<List<ConsultationDieteticien>> GetProchainesAsync(long utilisateurId)
        => await _ctx.Consultations
            .Where(c => c.UtilisateurId == utilisateurId && c.DateHeure > DateTime.Now)
            .OrderBy(c => c.DateHeure)
            .ToListAsync();

    public async Task<ConsultationDieteticien?> GetByIdAsync(long id)
        => await _ctx.Consultations.FindAsync(id);
}
